core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,226 @@
+   * context which is already closed.
+   *
+   * @throws java.lang.IllegalStateException if this context is not {@link #current()}.
+   * @return the {@link #current()} context after {@code this} has been detached.
+   */
+  // Discussion point: is the return useful? inconsistency with other fluent methods if we have them
+  public Context detach() {
+    throw new UnsupportedOperationException();
+  }
+
+
+  /**
+   * Attach this context to the thread and make it current, the previously current context
+   * will be restored when detach is called.
+   * @return
+   */
+  // Discussion point, make fluent?
+  public void attach() {
+    throw new UnsupportedOperationException();
+  }
+
+
+  /**
+   * Close this context and optionally provide a cause for the closure, all listeners are notified
+   * immediately on the calling thread
+   *
+   * @return {@code true} if this call closed the context and notified listeners, {@code false} if
+   *          the context was already closed.
+   */
+  public boolean close(@Nullable Throwable cause) {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Close the context normally.
+   */
+  @Override
+  // Discussion point: Cant return boolean here and be Closeable
+  public void close() {
+    close(null);
+  }
+
+  /**
+   * Has this context been closed.
+   */
+  public boolean isClosed() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Return the cause of the closure or {@code null} if context closed without a cause.
+   */
+  @Nullable
+  public Throwable cause() {
+    throw new UnsupportedOperationException();
+  }
+
+
+  public Runnable wrap(final Runnable r) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        attach();
+        try {
+          r.run();
+        } finally {
+          detach();
+        }
+      }
+    };
+  }
+
+  public <C> Callable<C> wrap(final Callable<C> c) {
+    return new Callable<C>() {
+      @Override
+      public C call() throws Exception {
+        attach();
+        try {
+          return c.call();
+        } finally {
+          detach();
+        }
+      }
+    };
+  }
+
+  public ExecutorService wrap(final ExecutorService e) {
+    // TODO
+    return null;
+  }
+
+  public ScheduledExecutorService wrap(final ScheduledExecutorService e) {
+    // TODO
+    return null;
+  }
+
+  // Discussion point, for Java 6+ would be nice to be able to wrap ForkJoinPool/Task
+  // and some of the newer stuff in Java8 too.
+
+  /**
+   * Create a {@link Key} with the given name.
+   */
+  public static <T> Key<T> key(String name) {
+    return new Key<T>(name);
+  }
+
+  /**
+   * Create a {@link Key} with the given name and default value.
+   */
+  public static <T> Key<T> keyWithDefault(String name, T defaultValue) {
+    return new Key<T>(name, defaultValue);
+  }
+
+  /**
+   *
+   * @param <T>
+   */
+  public static class Key<T> {
```

Comment:
Except that it's a `Key` that has a key and a value which is rather odd. If Context is essentially a map of names to values, then perhaps we should just follow the Java map naming conventions rather than inventing our own terminology.  `Entry`?